File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.io.*;
import java.util.*;

public class Main {

    public static boolean[][] visited;
    public static int[][] map;
    public static int N,a;
    public static int[][] dir = {{-1,-2},{-2,-1},{-2,1},{-1,2},{1,2},{2,1},{2,-1},{1,-2}};
    public static int Start_X, Start_Y, End_X, End_Y;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            a = Integer.parseInt(br.readLine());
            visited = new boolean[a][a];
            map = new int[a][a];

            st = new StringTokenizer(br.readLine()," ");
            Start_X = Integer.parseInt(st.nextToken());
            Start_Y = Integer.parseInt(st.nextToken());

            st = new StringTokenizer(br.readLine()," ");
            End_X = Integer.parseInt(st.nextToken());
            End_Y = Integer.parseInt(st.nextToken());

            bfs(new Var(Start_X,Start_Y));
            System.out.println(map[End_X][End_Y]);
        }
    }

    public static void bfs(Var d) {
        Queue<Var> q = new LinkedList<>();
        q.offer(d);
        visited[d.x][d.y] = true;

        while(!q.isEmpty()) {
            Var b = q.poll();
            int x = b.x;
            int y = b.y;

            if(x == End_X && y == End_Y) {
                break;
            }

            for (int i = 0; i < 8; i++) {
                int nx = x + dir[i][0];
                int ny = y + dir[i][1];

                if(nx >=0 && ny >=0 && nx < a && ny < a) {
                    if(!visited[nx][ny]) {
                        visited[nx][ny] = true;
                        map[nx][ny] = map[x][y] + 1;
                        q.offer(new Var(nx,ny));
                    }
                }
            }
        }
    }

    static class Var {
        int x;
        int y;

        public Var(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque
sys.setrecursionlimit(10**6)

T = int(sys.stdin.readline())

dx = [-2, -2, -1, -1, 1, 1, 2, 2]
dy = [1, -1, 2, -2, 2, -2, -1, 1]

def bfs(x1, y1, x2, y2, I):
    
    visited = [[0] * I for _ in range(I)]
    
    visited[x1][y1] = 1
    queue = deque([(x1, y1)])
   
    
    while queue:
        x, y = queue.popleft()
        
        if x == x2 and y == y2:
            return visited[x][y] - 1
    
        for i in range(8):
            nx = x + dx[i]
            ny = y + dy[i]
        
            if 0 <= nx < I and 0 <= ny < I and not visited[nx][ny]:
                visited[nx][ny] = visited[x][y] + 1
                queue.append((nx, ny))
    return 0

for _ in range(T):
    
    I = int(sys.stdin.readline())
    
    x1, y1 = map(int, sys.stdin.readline().split())
    x2, y2 = map(int, sys.stdin.readline().split())
    
    print(bfs(x1, y1, x2, y2, I))